File: lib/src/domain/user/entities/user.entity.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:linko/src/domain/chat/entities/chat.entity.dart';
import 'package:linko/src/domain/company/entities/company.entity.dart';
import 'package:linko/src/domain/company/entities/tag.entity.dart';
import 'package:linko/src/domain/user/entities/favorites.entity.dart';

import '../../auth/entities/token.entity.dart';

part 'user.entity.g.dart';

@JsonSerializable()
class UserEntity extends Equatable {
  int id;
  String? UID;
  String? avatar;
  String? fullName;
  String phoneNumber;
  String role;
  String language;
  String theme;
  bool isActive;
  List<TokenEntity>? tokens;

  List<TagEntity>? tags;
  List<CompanyEntity>? companies;
  List<FavoriteEntity>? favorites;
  List<ChatEntity>? chatSent;
  List<ChatEntity>? chatReceived;

  UserEntity(
      {required this.id,
      this.UID,
      this.avatar,
      this.fullName,
      required this.phoneNumber,
      this.role = 'USER',
      this.language = 'en-us',
      this.theme = 'light',
      this.isActive = true,
      this.tokens,
      this.tags,
      this.companies,
      this.favorites,
      this.chatSent,
      this.chatReceived});

  @override
  List<Object?> get props => [
        id,
        UID,
        avatar,
        fullName,
        phoneNumber,
        role,
        language,
        theme,
        isActive,
        tokens,
        tags,
        companies,
        favorites,
        chatSent,
        chatReceived
      ];

  factory UserEntity.fromJson(Map<String, dynamic> json) =>
      _$UserEntityFromJson(json);

  Map<String, dynamic> toJson() => _$UserEntityToJson(this);

  dynamic get(String propName) {
    switch (propName) {
      default:
        return fullName;
    }
  }

  static Future<List<UserEntity>> loadUsers(l) async {
    return List<UserEntity>.from(l.map((model) => UserEntity.fromJson(model)));
  }
}

File: src/domain/user/entity/user.entity.ts
import { Chat } from "../../chat/entity/chat.entity";
import { Company } from "../../company/entity/company.entity";
import { Tag } from "../../company/entity/tag.entity";
import { Token } from "./token.entity";
import { Favorite } from "./favorite.entity";

export class User {
  id?: number;

  UID?: string;

  avatar?: string;

  fullName?: string;

  phoneNumber?: string;

  role?: string;

  language?: string;

  theme?: string;

  isActive?: boolean;

  tokens?: Token[];

  tags?: Tag[];

  companies?: Company[];

  favorites?: Favorite[];

  chatSent?: Chat[];

  chatReceived?: Chat[];

  constructor(data: any) {
    this.id = Number(data?.id);
    this.fullName = String(data?.fullName);
    this.avatar = String(data?.avatar);
    this.UID = String(data?.UID);
    this.phoneNumber = String(data?.phoneNumber);
    this.theme = String(data?.theme);
    this.role = String(data?.role);
    this.language = String(data?.language);
    this.isActive = Boolean(data?.isActive);

    if (data?.tags != undefined && data?.tags != null)
      this.tokens = (data?.tokens).map((t: any) => new Token(t));

    // if (data?.tags != undefined && data?.tags != null)
    //   this.tags = (data?.tags).map((t: any) => new Tag(t));

    // if (data?.tags != undefined && data?.tags != null)
    //   this.companies = (data?.companies).map((t: any) => new Company(t));

    // if (data?.tags != undefined && data?.tags != null)
    //   this.favorites = (data?.favorites).map((t: any) => new Favorite(t));

    // if (data?.tags != undefined && data?.tags != null)
    //   this.chatSent = (data?.chatSent).map((t: any) => new Chat(t));

    // if (data?.tags != undefined && data?.tags != null)
    //   this.chatReceived = (data?.chatReceived).map((t: any) => new Chat(t));
  }
}
